### PR TITLE
ci: ensure trunk is always green

### LINKED ISSUE #9945: Merge Conflicts for `CHANGELOG_PENDING`
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

We get many merge conflicts when updating changelog pending. We should fix this.

## Proposed Solutions

1. We could create a changelog folder and concatenate changes together when updating `CHANGELOG.md`.
2. Structured YAML to provide easier merge resolution.

### LINKED ISSUE #10091: Reduce PR blocking and keep trunk green with merge queues
We would like to prevent false results in our testing system and increase CI velocity by moving many of our CI checks to a step _between_ clicking merge PR and the result being placed onto the master branch.

This could occur via the [bors bot](https://bors.tech/) or likely later this year, [GitHub merge queues](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue).

### PR DESCRIPTION
**Pull request by AaronFriel**
Implement `bors` based merge queues and a rework of our CI pipeline with an emphasis on quick, actionable results and comprehensive pre-merge testing. Ensures that the merge target - `master` -remains green.

A theme of this PR is treating GitHub workflows more like functions, thanks to new capabilities to split workflows and separate concerns. Each reusable workflow begins with a relatively common header of inputs: `ref`, `version`, sometimes `platform`, a list of language versions to use.

Another theme of this PR is not running work on the target branch, only running *creative* operations, those that create or modify artifacts, on the bors queue branches, and fully reusing artifacts from job stage to job stage.

This pattern is reused throughout the new jobs to ensure that, e.g.: whether on a merge, on a staging PR, or on a `/run-acceptance-tests` command, we use the correct branch and uniformly select the version. (The `version` parameter replaces `versions.txt`.)

Testing the last stage (release) is a bit difficult, as the final steps of the release workflow are not validated to accept alpha versions. That said, I'm confident the process works because the [entire workflow](https://github.com/pulumi/pulumi/actions/runs/2982880324) succeeded. See:
- https://github.com/pulumi/pulumi/issues/10616
- https://github.com/pulumi/pulumi/issues/10612

However, the package managers and Git (aka Go modules package manager) support alphas.

Implements the changelog tool in this repo, after taking feedback on a PR testing it here:
- https://github.com/pulumi/watchutil-rs/pull/30

Resolves #9945 
Resolves #10091

### CODE REVIEW

File: .goreleaser.yml [hunk 1]
```diff
@@ -19,40 +14,38 @@ builds:
...
+  main: ./cmd/pulumi
+  gobinary: ../scripts/go-wrapper.sh
+  env:
+  - CGO_ENABLED=0
+  - GO111MODULE=on
   goos: ['linux', 'darwin', 'windows']
   goarch: ['amd64', 'arm64']
   goamd64: ['v1']
-  ignore:
-    - goos: windows
```
Comment (t0yv0) [inline .goreleaser.yml:26]: Do we build windows/arm64 now?

File: .goreleaser.yml [hunk 2]
```diff
@@ -19,40 +14,38 @@ builds:
 - &pulumibin
   id: pulumi
   binary: pulumi
+  dir: pkg
```
Comment (t0yv0) [inline .goreleaser.yml:17]: Suggest maybe dropping these sections, if we are handling this elsewhere outside goreleaser now:

- brews (opens a pr in pulumi/homebrew-tap)
- release (opens GH rel + discussions)
- blobs (uploads to S3)

File: .github/workflows/release.pub [hunk 3]
```diff
@@ -0,0 +1,2 @@
+untrusted comment: minisign public key: DE23EFE715137751
```
Comment (t0yv0) [inline .github/workflows/release.pub:1]: What is this from/for?

File: scripts/get-version [hunk 4]
```diff
@@ -8,7 +8,7 @@ fi
 
 ## PLEASE NOTE: This script is used primarily for use with homebrew and the "make brew" build target
 ## the logic to determine the version of pulumi at build time is now controlled via
-## https://github.com/pulumi/pulumictl and the `pulumictl get version` command
+## scripts/pulumi-version.sh and the PULUMI_VERSION environment variable.
```
Comment (t0yv0) [inline scripts/get-version:11]: There's something going on in https://github.com/pulumi/pulumi/commits/master/scripts/get-version

It seems that we attempted to delete this script in favor of pulumictl to do versioning, then partially reverted that, but I can't find the reason. Looks like pulumictl version detection does not satisfy all the reqs here? 

Can we shed some light on this? 

Are we going to remove pulumictl now and maintain this script instead?

Would it possibly be better to patch pulumictl for the missing functionality? Happy to help hacking on this.

File: .github/workflows/release-pr.yml [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            cd pkg
+            go get -u "github.com/pulumi/pulumi/sdk/v3@v${PULUMI_VERSION}"
+          )
+          make tidy
+          git add -A
+          git commit -m "Update dependencies"
+
+          # Publish pkg module on another tag.
+          git tag "pkg/v${PULUMI_VERSION}"
+          git push origin "pkg/v${PULUMI_VERSION}"
```
Comment (t0yv0) [inline .github/workflows/release-pr.yml:82]: Curious how this works in go tooling. If we push pkg/v3.100.0 tag, but on a branch that is not the main branch of a repository, will `go` tooling consider that published, e.g. resolve go.mod references to v3.100.0 to that tag even if it is not on master? Checking this out quickly.

Comment (t0yv0) [reply .github/workflows/release-pr.yml:82]: Checked that! Never mind, all good. `go mod` does not resolve a new tag until it's on main branch.

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -1,461 +1,192 @@
...
     strategy:
+      fail-fast: false
       matrix:
-        go-version: [1.17.x]
-        python-version: [3.9.x]
-        dotnet-version: [6.0.x]
-        node-version: [14.x]
-    runs-on: ubuntu-latest
+        job:
+          - name: Templates Smoke Test
```
Comment (t0yv0) [inline .github/workflows/release.yml:164]: Has this action run? How do we test run this logic, or perhaps a subset of it?  

I'm assuming using a pre-release release through official channels is not an option.

We could try to test-drive against `<forked>/examples` etc, maybe comment out winget-deploy and choco-deploy, whatever is not easy to run in forked mode.

File: .goreleaser.yml (hunk 1 above)
Comment (AaronFriel) [reply .goreleaser.yml:26]: We can & should - it's a platform Go supports and Windows, Windows Server support them as platforms.

File: .github/workflows/release.pub (hunk 3 above)
Comment (AaronFriel) [reply .github/workflows/release.pub:1]: A signing key for signing checksums - I use it on pulumi-watch. I'll remove it, as we need to discuss this as a team. It's unused (the sign job is disabled.)

File: scripts/get-version (hunk 4 above)
Comment (AaronFriel) [reply scripts/get-version:11]: The .version file is the canonical "release" version we are targeting in a repository, gives us control to edit the file to `3.x.1` if we wish to make a patch release.

`pulumictl get version` is based around git tags, tangentially this requires a deep clone resulting in slower checkouts, but more importantly, we aren't using git tags. All release builds will be on an untagged branch:

```
pulumictl get version --version-prefix 3.5.0
3.5.0-alpha.1662663630+52dcaaf7
```

I could patch `pulumictl get version` to take a literal version, but the shell script is much easier to understand at much lower complexity if we should like to change it.

File: .github/workflows/release.yml (hunk 6 above)
Comment (AaronFriel) [reply .github/workflows/release.yml:164]: I'm not sure - I think triggering examples, templates, etc tests on PRs and during merge queues is out of scope.

It has definitely run: https://github.com/pulumi/pulumi/runs/8166609012?check_suite_focus=true

Post-release, would you like to see an issue created to:
* Fixing examples & templates tests to pass reliably.
* For those & others, enabling testing of our post-release actions *during* the merge queue workflow? That is, with a snapshot build, or on a cron?

File: .goreleaser.yml (hunk 2 above)
Comment (AaronFriel) [reply .goreleaser.yml:17]: Good call, and on that note I need to add the GH discussion to a post-release action.

File: scripts/get-version (hunk 4 above)
Comment (t0yv0) [reply scripts/get-version:11]: Whether the team prefers to maintain bash or Go tooling - we can figure out later. This should not be blocking the PR.

File: .github/workflows/release.yml (hunk 6 above)
Comment (t0yv0) [reply .github/workflows/release.yml:164]: I think we converged on trying this out in-place using a real point release of pulumi/pulumi  - my biggest concern was just that we have some logic in the yamls here that might have never been executed anywhere.

These issues sound helpful yes, in the area/build backlog. We can later pick up a subset of examples/tests to verify against.

**Review by t0yv0 — APPROVED** [review]
LGTM

**Comment by AaronFriel** [comment]
bors merge

**Comment by AaronFriel** [comment]
bors retry

**Comment by AaronFriel** [comment]
bors cancel

**Comment by AaronFriel** [comment]
bors retry

**Comment by AaronFriel** [comment]
bors merge

**Comment by AaronFriel** [comment]
bors retry

**Pull request merged**